File: python/src/br_validator/ie/_rj.py
"""RJ: verified against the official SEFAZ-RJ "Roteiro de Crítica da
Inscrição Estadual" (sintegra.gov.br/Cad_Estados/cad_RJ.html) for the
check-digit rule, and cross-checked against a worked example (99.999.99-3)
for the weights, which the Sintegra page itself does not enumerate.
"""

import re

from .._shared import remove_non_digits
from ._mod11 import weighted_sum

_ALLOWED_CHARS = re.compile(r"^[\d.\-\s]+$")
_WEIGHTS = [2, 7, 6, 5, 4, 3, 2]


def normalize(value: str) -> str:
    return remove_non_digits(value)


def _calculate_check_digit(v: str) -> int:
    total = weighted_sum(v, _WEIGHTS)
    remainder = total % 11

    if remainder <= 1:
        return 0

    return 11 - remainder


def is_valid(value: str) -> bool:
    if not _ALLOWED_CHARS.match(value):
        return False

    v = normalize(value)

    if len(v) != 8:
        return False

    return _calculate_check_digit(v) == int(v[7])


def format(value: str) -> str:
    v = normalize(value)

    if len(v) != 8:
        return value

    return f"{v[:2]}.{v[2:5]}.{v[5:7]}-{v[7:8]}"

File: ruby/lib/br_validator/ie/rj.rb
# frozen_string_literal: true

require_relative "../shared"
require_relative "mod11"

module BrValidator
  module Ie
    # RJ: verified against the official SEFAZ-RJ "Roteiro de Crítica da
    # Inscrição Estadual" (sintegra.gov.br/Cad_Estados/cad_RJ.html) for the
    # check-digit rule, and cross-checked against a worked example
    # (99.999.99-3) for the weights, which the Sintegra page itself does
    # not enumerate.
    module Rj
      module_function

      ALLOWED_CHARS = /\A[\d.\-\s]+\z/.freeze
      WEIGHTS = [2, 7, 6, 5, 4, 3, 2].freeze
      private_constant :ALLOWED_CHARS, :WEIGHTS

      def normalize(value)
        Shared.remove_non_digits(value)
      end

      def calculate_check_digit(v)
        sum = Mod11.weighted_sum(v, WEIGHTS)
        remainder = sum % 11

        return 0 if remainder <= 1

        11 - remainder
      end
      private_class_method :calculate_check_digit

      def is_valid?(value)
        return false unless value.match?(ALLOWED_CHARS)

        v = normalize(value)

        return false unless v.length == 8

        calculate_check_digit(v) == v[7].to_i
      end

      def format(value)
        v = normalize(value)

        return value unless v.length == 8

        "#{v[0...2]}.#{v[2...5]}.#{v[5...7]}-#{v[7...8]}"
      end
    end
  end
end
